src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -290,6 +296,97 @@ class SegmentedAllocator : public ExtentAllocator {
   Cache& cache;
 };
 
+/*
+ * RandomBlock allocator
+ *
+ */
+class RBAllocator : public ExtentAllocator {
+  class Writer : public ExtentOolWriter {
+  public:
+    Writer(
+      RandomBlockManager& rbm,
+      LBAManager& lba_manager,
+      Cache& cache)
+      : rbm(rbm),
+        lba_manager(lba_manager),
+	cache(cache)
+    {}
+    Writer(Writer &&) = default;
+
+    write_iertr::future<> write(
+      Transaction& t,
+      std::list<LogicalCachedExtentRef>& extent) final;
+    stop_ertr::future<> stop() final {
+      return writer_guard.close().then([] {
+	  return stop_ertr::now();
+      });
+    }
+  private:
+    using update_lba_mapping_iertr = LBAManager::update_le_mapping_iertr;
+    using finish_record_iertr = update_lba_mapping_iertr;
+    using finish_record_ret = finish_record_iertr::future<>;
+    finish_record_ret finish_write(
+      Transaction& t,
+      ool_record_t& record);
+
+    write_iertr::future<> _write(
+      Transaction& t,
+      ool_record_t& record);
+
+    using extents_to_write_t = std::vector<LogicalCachedExtentRef>;
+    void add_extent_to_write(
+      ool_record_t&,
+      LogicalCachedExtentRef& extent);
+
+    RandomBlockManager& rbm;
+    LBAManager& lba_manager;
+    seastar::gate writer_guard;
+    Cache& cache;
+  };
+public:
+  RBAllocator(
+    RandomBlockManager& rbm,
+    LBAManager& lba_manager,
+    Cache& cache);
+
+  Writer &get_writer(placement_hint_t hint) {
+    return writers[std::rand() % writers.size()];
+  }
+
+  alloc_paddr_iertr::future<> alloc_ool_extents_paddr(
+    Transaction& t,
+    std::list<LogicalCachedExtentRef>& extents) final {
+    return seastar::do_with(
+      std::map<Writer*, std::list<LogicalCachedExtentRef>>(),
+      [this, extents=std::move(extents), &t](auto& alloc_map) {
+      for (auto& extent : extents) {
+        auto writer = &(get_writer(extent->hint));
+        alloc_map[writer].emplace_back(extent);
+      }
+      return trans_intr::do_for_each(alloc_map, [&t](auto& p) {
+        auto writer = p.first;
+        auto& extents_to_persist = p.second;
+	return writer->write(t, extents_to_persist);
```

Comment:
I think, for random block devices, we don't have to do actual ool writes at the time of delayed extent allocation. The reason why we have to do actual ool writes for segment devices is that, for segment devices, we can only confirm the paddr_t that an extent will be written to after that extent is written. Random block devices don't have this issue, right?

I think maybe we should add a new method like "do_ool_writes()" to EPM, and postpone random block devices' actual ool writes(the invocation of do_ool_writes) to the journal commit phase. With this approach, we can avoid the write amplifications caused by transaction invalidation that happened after the delayed extent allocation. However, this comes at the price of reduced concurrency of ool writes, as only one transaction's ool writes can be done at any time.

@myoungwon @athanatos What do you think?:-)

Reply:
> But, in case of a large block (large size record), according to RBM design, we don't have to write the large block to the segment device (CBJournal here), rather we want to write the large block to the RBM directly.

Um, I meant that we just postpone the ool write to the journal commit phase, but still write to RBM instead of CBJournal. In other words, maybe we should give journals the ability to write to other segments/RBMs.

Reply:
I got it. The postpone is ok. But, I'm not sure that current journal needs to take a role of writing both segment and RBM. Maybe we need a new absraction that can cover both Segment and RBM?

I have one question. When the transaction invalidation happens? Even the invalidation occurs, the write procedure---write for RBM should be preceded before appending record to CBjournal---would not be changes. The invalidation can happen between RBM write and CBjournal write?

Reply:
> Maybe we need a new absraction that can cover both Segment and RBM?

Yes, that's also my belief:-)

> I have one question. When the transaction invalidation happens? Even the invalidation occurs, the write procedure---write for RBM should be preceded before appending record to CBjournal---would not be changes. The invalidation can happen between RBM write and CBjournal write?

I think a transaction can be invalidated at any time before it enters the `WritePipeline::prepare` phase, so as long as we do write for RBM after entering that phase, there won't be invalidation between RBM write and CBjournal write.

@athanatos Am I right?:-)

Reply:
Sorry, I've been a bit busy, I'll try to catch up on this tomorrow.